src/atomics/seqcst.md

```diff
@@ -0,0 +1,445 @@
+const ORDERING: atomic::Ordering = atomic::Ordering::Relaxed;
+
+static X: AtomicBool = AtomicBool::new(false);
+static Y: AtomicBool = AtomicBool::new(false);
+
+let a = thread::spawn(|| { X.store(true, ORDERING) });
+let b = thread::spawn(|| { Y.store(true, ORDERING) });
+let c = thread::spawn(|| { while !X.load(ORDERING) {} Y.load(ORDERING) });
+let d = thread::spawn(|| { while !Y.load(ORDERING) {} X.load(ORDERING) });
+
+let a = a.join().unwrap();
+let b = b.join().unwrap();
+let c = c.join().unwrap();
+let d = d.join().unwrap();
+
+# return;
+// This assert is allowed to fail.
+assert!(c || d);
+```
+
+The basic setup of this code, for all of its possible executions, looks like
+this:
+
+```text
+     a        static X         c               d        static Y         b
+╭─────────╮   ┌───────┐   ╭─────────╮     ╭─────────╮   ┌───────┐   ╭─────────╮
+│ store X ├─┐ │ false │ ┌─┤ load X  │     │ load Y  ├─┐ │ false │ ┌─┤ store Y │
+╰─────────╯ │ └───────┘ │ ╰────╥────╯     ╰────╥────╯ │ └───────┘ │ ╰─────────╯
+            └─┬───────┐ │ ╭────⇓────╮     ╭────⇓────╮ │ ┌───────┬─┘
+              │ true  ├─┘ │ load Y  ├─? ?─┤ load X  │ └─┤ true  │
+              └───────┘   ╰─────────╯     ╰─────────╯   └───────┘
+```
+
+In other words, `a` and `b` are guaranteed to, at some point, store `true` into
+`X` and `Y` respectively, and `c` and `d` are guaranteed to, at some point, load
+those values of `true` from `X` and `Y` (there could also be an arbitrary number
+of loads of `false` by `c` and `d`, but they’ve been omitted since they don’t
+actually affect the execution at all). The question now is when `c` and `d` load
+from `Y` and `X` respectively, is it possible for them _both_ to load `false`?
+
+And looking at this diagram, there’s absolutely no reason why not. There isn’t
+even a single arrow connecting the left and right hand sides so far, so the load
+has no coherence-based restrictions on which value it is allowed to pick — and
+this goes for both sides equally, so we could end up with an execution like
+this:
+
+```text
+     a        static X         c             d        static Y         b
+╭─────────╮   ┌───────┐   ╭─────────╮   ╭─────────╮   ┌───────┐   ╭─────────╮
+│ store X ├─┐ │ false ├┐ ┌┤ load X  │   │ load Y  ├┐ ┌┤ false │ ┌─┤ store Y │
+╰─────────╯ │ └───────┘│ │╰────╥────╯   ╰────╥────╯│ │└───────┘ │ ╰─────────╯
+            └─┬───────┐└─│─────║──────┐┌─────║─────│─┘┌───────┬─┘
+              │ true  ├──┘╭────⇓────╮┌─┘╭────⇓────╮└──┤ true  │
+              └───────┘   │ load Y  ├┘└─┤ load X  │   └───────┘
+                          ╰─────────╯   ╰─────────╯
+```
+
+Which results in a failed assert. This execution is brought about because the
+model of separate modification orders means that there is no relative ordering
+between `X` and `Y` being changed, and so each thread is allowed to “see” either
+order. However, some algorithms will require a globally agreed-upon ordering,
+and this is where `SeqCst` can come in useful.
```

Comment:
I'd be interested to know which algorithms actually require SeqCst. The leading motivation for proposing algorithms under SeqCst is probably that it's easier to reason about, not that a Single Total Order is necessary (read-latest may be needed but this can be done with idempotent RMW).